# Data Scientist VS Data analyst

## Plan of Attack
Explore the structure of the data


# Download the data 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
#display 
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
import warnings
warnings.filterwarnings(action='ignore')

import matplotlib.pyplot as plt #likley won't be used much as i'm experimenting with plotly 
import plotly.graph_objects as go #you will be learning how go and px work with me! 
import plotly.express as px

In [2]:
import pandas as pd
import numpy as np
from sklearn.utils import shuffle
from IPython.core.display import display, HTML, Javascript
from string import Template
import json, random
import IPython.display
from plotly.offline import init_notebook_mode, iplot
from plotly import subplots
import plotly.figure_factory as ff
import plotly as py
import plotly.graph_objects as go
init_notebook_mode(connected=True)

df=pd.read_csv('kaggle_survey_2020_responses.csv')
df = df.drop(df.index[0])
questions = df.columns

# Defining all our palette colours.
primary_blue = "#496595"
primary_blue2 = "#85a1c1"
primary_blue3 = "#3f4d63"
primary_grey = "#c6ccd8"
primary_black = "#202022"
primary_bgcolor = "#f4f0ea"

# "coffee" pallette turqoise-gold.
f1 = "#a2885e"
f2 = "#e9cf87"
f3 = "#f1efd9"
f4 = "#8eb3aa"
f5 = "#235f83"
f6 = "#b4cde3"

# Take a quick look at the Data Structure

In [3]:
df.head()

Time from Start to Finish (seconds)     Q1   Q2                        Q3  \
1                                1838  35-39  Man                  Colombia   
2                              289287  30-34  Man  United States of America   
3                                 860  35-39  Man                 Argentina   
4                                 507  30-34  Man  United States of America   
5                                  78  30-34  Man                     Japan   

                  Q4                 Q5           Q6 Q7_Part_1 Q7_Part_2  \
1    Doctoral degree            Student   5-10 years    Python         R   
2    Master’s degree      Data Engineer   5-10 years    Python         R   
3  Bachelor’s degree  Software Engineer  10-20 years       NaN       NaN   
4    Master’s degree     Data Scientist   5-10 years    Python       NaN   
5    Master’s degree  Software Engineer    3-5 years    Python       NaN   

  Q7_Part_3 Q7_Part_4 Q7_Part_5 Q7_Part_6   Q7_Part_7 Q7_Part_8 Q7_Part_9  \
1       SQL         C       NaN       NaN  Javascript       NaN       NaN   
2       SQL       NaN       NaN       NaN         NaN       NaN       NaN   
3       NaN       NaN       NaN      Java  Javascript       NaN       NaN   
4       SQL       NaN       NaN       NaN         NaN       NaN       NaN   
5       NaN       NaN       NaN       NaN         NaN       NaN       NaN   

  Q7_Part_10 Q7_Part_11 Q7_Part_12 Q7_OTHER      Q8  \
1        NaN     MATLAB        NaN    Other  Python   
2        NaN        NaN        NaN      NaN  Python   
3       Bash        NaN        NaN      NaN       R   
4       Bash        NaN        NaN      NaN  Python   
5        NaN        NaN        NaN      NaN  Python   

                                       Q9_Part_1 Q9_Part_2      Q9_Part_3  \
1  Jupyter (JupyterLab, Jupyter Notebooks, etc)        NaN            NaN   
2                                            NaN       NaN  Visual Studio   
3                                            NaN       NaN            NaN   
4                                            NaN       NaN            NaN   
5                                            NaN       NaN            NaN   

                     Q9_Part_4  Q9_Part_5   Q9_Part_6      Q9_Part_7  \
1  Visual Studio Code (VSCode)        NaN    Spyder              NaN   
2                          NaN   PyCharm          NaN            NaN   
3  Visual Studio Code (VSCode)        NaN         NaN    Notepad++     
4                          NaN   PyCharm          NaN            NaN   
5                          NaN        NaN         NaN            NaN   

          Q9_Part_8        Q9_Part_9 Q9_Part_10 Q9_Part_11 Q9_OTHER  \
1               NaN              NaN        NaN        NaN      NaN   
2    Sublime Text                NaN        NaN        NaN      NaN   
3    Sublime Text      Vim / Emacs          NaN        NaN      NaN   
4               NaN              NaN        NaN        NaN      NaN   
5               NaN              NaN        NaN        NaN      NaN   

          Q10_Part_1       Q10_Part_2 Q10_Part_3 Q10_Part_4 Q10_Part_5  \
1   Kaggle Notebooks  Colab Notebooks        NaN        NaN        NaN   
2                NaN  Colab Notebooks        NaN        NaN        NaN   
3                NaN              NaN        NaN        NaN        NaN   
4                NaN              NaN        NaN        NaN        NaN   
5                NaN              NaN        NaN        NaN        NaN   

  Q10_Part_6 Q10_Part_7 Q10_Part_8 Q10_Part_9 Q10_Part_10 Q10_Part_11  \
1        NaN        NaN        NaN        NaN         NaN         NaN   
2        NaN        NaN        NaN        NaN         NaN         NaN   
3        NaN        NaN        NaN        NaN         NaN         NaN   
4        NaN        NaN        NaN        NaN         NaN         NaN   
5        NaN        NaN        NaN        NaN         NaN         NaN   

  Q10_Part_12 Q10_Part_13 Q10_OTHER  \
1         NaN         NaN     

We have question with only one choice and question with multiple answers

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20036 entries, 1 to 20036
Columns: 355 entries, Time from Start to Finish (seconds) to Q35_B_OTHER
dtypes: object(355)
memory usage: 54.4+ MB


There are 20037 in the dataset and 355 columns, that is fairly small by Machine Learning standards

In [5]:
#Missing value
df.isnull().sum()/df.shape[0]

Time from Start to Finish (seconds)    0.000000
Q1                                     0.000000
Q2                                     0.000000
Q3                                     0.000000
Q4                                     0.023308
Q5                                     0.037882
Q6                                     0.045718
Q7_Part_1                              0.224895
Q7_Part_2                              0.786534
Q7_Part_3                              0.623927
Q7_Part_4                              0.834548
Q7_Part_5                              0.808994
Q7_Part_6                              0.831952
Q7_Part_7                              0.850519
Q7_Part_8                              0.986924
Q7_Part_9                              0.990118
Q7_Part_10                             0.911360
Q7_Part_11                             0.889349
Q7_Part_12                             0.989719
Q7_OTHER                               0.902925
Q8                                     0

We can see that we can find a lot of null value but we can think this is due to the fact that the questions have multiple answers

In [6]:
df.describe()

Time from Start to Finish (seconds)     Q1     Q2     Q3  \
count                                 20036  20036  20036  20036   
unique                                 5167     11      5     55   
top                                     565  25-29    Man  India   
freq                                     34   4011  15789   5851   

                     Q4       Q5         Q6 Q7_Part_1 Q7_Part_2 Q7_Part_3  \
count             19569    19277      19120     15530      4277      7535   
unique                7       13          7         1         1         1   
top     Master’s degree  Student  3-5 years    Python         R       SQL   
freq               7859     5171       4546     15530      4277      7535   

       Q7_Part_4 Q7_Part_5 Q7_Part_6   Q7_Part_7 Q7_Part_8 Q7_Part_9  \
count       3315      3827      3367        2995       262       198   
unique         1         1         1           1         1         1   
top            C       C++      Java  Javascript     Julia     Swift   
freq        3315      3827      3367        2995       262       198   

       Q7_Part_10 Q7_Part_11 Q7_Part_12 Q7_OTHER      Q8  \
count        1776       2217        206     1945   17821   
unique          1          1          1        1      13   
top          Bash     MATLAB       None    Other  Python   
freq         1776       2217        206     1945   14241   

                                            Q9_Part_1  Q9_Part_2  \
count                                           11211       3826   
unique                                              1          1   
top     Jupyter (JupyterLab, Jupyter Notebooks, etc)    RStudio    
freq                                            11211       3826   

            Q9_Part_3                    Q9_Part_4  Q9_Part_5   Q9_Part_6  \
count            2445                         5873       5099        3290   
unique              1                            1          1           1   
top     Visual Studio  Visual Studio Code (VSCode)   PyCharm     Spyder     
freq             2445                         5873       5099        3290   

            Q9_Part_7         Q9_Part_8        Q9_Part_9 Q9_Part_10  \
count            3132              2452             1502       1604   
unique              1                 1                1          1   
top       Notepad++      Sublime Text      Vim / Emacs      MATLAB    
freq             3132              2452             1502       1604   

       Q9_Part_11 Q9_OTHER         Q10_Part_1       Q10_Part_2  \
count         386     1162               5992             6329   
unique          1        1                  1                1   
top          None    Other   Kaggle Notebooks  Colab Notebooks   
freq          386     1162               5992             6329   

             Q10_Part_3               Q10_Part_4             Q10_Part_5  \
count               857                      180                   2072   
unique                1                        1                      1   
top     Azure Notebooks   Paperspace / Gradient    Binder / JupyterHub    
freq                857                      180                   2072   

          Q10_Part_6           Q10_Part_7                 Q10_Part_8  \
count            105                  846                        497   
unique             1                    1                          1   
top      Code Ocean    IBM Watson Studio    Amazon Sagemaker Studio    
freq             105                  846                        497   

                    Q10_Part_9                          Q10_Part_10  \
count                      245                                 1218   
unique                       1                                    1   
top      Amazon EMR Notebooks   Google Cloud AI Platform Notebooks    
freq                       245                                 1218   

                           Q10_Part_11                           Q10_Part_12  \
count                           

#### Main Observations

>The majority of the respondants are from India for more than 40% has a master degree, writting code for 3-5 years, use Python, R or SQL on regular basis and reccomend to learn Python to get into Data science field for 70%. 
They work on Juputer more than 50% but Visual Studio Code and Pycharm are very popular in our our dataset. We can see that Rstudio is only use by less than 20% of the respondant which is quite supprising and could make us imagine how Python starting to be dominant is the data field. Collab and Kaggle otebook are the most used and more than 66% of the respondents use a personal computer or laptop to work and 40% work with a GPU.

>Matplotlib (used by 60%) and Seaborn(60%) are the most popular visualization libraries. To talk about years of experiences in Machine Learning  30% has under 1 year experience which could be explained by 2 phenomenons that will confirm in the data exploration, hypothesis one a lot of respondant start the data science very recently second hypothese a lot of respondent are students and enroll at the University or College in Data.

>Let's talk about the most popular framework whithout supprise Skikit-Learn is the most common and 50% of the respondant use it followedby Keras and Tensorflow with more than 30%. Linear and Logistic Regression(50%) are are the msot well known with Random Forest




# Purpose of the study: Understand the difference between data Scientist and Data Analsyt 
In our study we will be interested in only to the data_scientist and Data Analyst we will try to asnwer to few questions=
### Question 1: Is there a significative difference in the graduation level ? 
### Question 2: Do the Data Scientist use the same programmation laguage ?
### QUestion 3: Do they both use Machine Learning Method an wich librairies ?
### Question 4: What are the main activities for their job roles 


# Prepare our dataset 

In [7]:
df.head()

Time from Start to Finish (seconds)     Q1   Q2                        Q3  \
1                                1838  35-39  Man                  Colombia   
2                              289287  30-34  Man  United States of America   
3                                 860  35-39  Man                 Argentina   
4                                 507  30-34  Man  United States of America   
5                                  78  30-34  Man                     Japan   

                  Q4                 Q5           Q6 Q7_Part_1 Q7_Part_2  \
1    Doctoral degree            Student   5-10 years    Python         R   
2    Master’s degree      Data Engineer   5-10 years    Python         R   
3  Bachelor’s degree  Software Engineer  10-20 years       NaN       NaN   
4    Master’s degree     Data Scientist   5-10 years    Python       NaN   
5    Master’s degree  Software Engineer    3-5 years    Python       NaN   

  Q7_Part_3 Q7_Part_4 Q7_Part_5 Q7_Part_6   Q7_Part_7 Q7_Part_8 Q7_Part_9  \
1       SQL         C       NaN       NaN  Javascript       NaN       NaN   
2       SQL       NaN       NaN       NaN         NaN       NaN       NaN   
3       NaN       NaN       NaN      Java  Javascript       NaN       NaN   
4       SQL       NaN       NaN       NaN         NaN       NaN       NaN   
5       NaN       NaN       NaN       NaN         NaN       NaN       NaN   

  Q7_Part_10 Q7_Part_11 Q7_Part_12 Q7_OTHER      Q8  \
1        NaN     MATLAB        NaN    Other  Python   
2        NaN        NaN        NaN      NaN  Python   
3       Bash        NaN        NaN      NaN       R   
4       Bash        NaN        NaN      NaN  Python   
5        NaN        NaN        NaN      NaN  Python   

                                       Q9_Part_1 Q9_Part_2      Q9_Part_3  \
1  Jupyter (JupyterLab, Jupyter Notebooks, etc)        NaN            NaN   
2                                            NaN       NaN  Visual Studio   
3                                            NaN       NaN            NaN   
4                                            NaN       NaN            NaN   
5                                            NaN       NaN            NaN   

                     Q9_Part_4  Q9_Part_5   Q9_Part_6      Q9_Part_7  \
1  Visual Studio Code (VSCode)        NaN    Spyder              NaN   
2                          NaN   PyCharm          NaN            NaN   
3  Visual Studio Code (VSCode)        NaN         NaN    Notepad++     
4                          NaN   PyCharm          NaN            NaN   
5                          NaN        NaN         NaN            NaN   

          Q9_Part_8        Q9_Part_9 Q9_Part_10 Q9_Part_11 Q9_OTHER  \
1               NaN              NaN        NaN        NaN      NaN   
2    Sublime Text                NaN        NaN        NaN      NaN   
3    Sublime Text      Vim / Emacs          NaN        NaN      NaN   
4               NaN              NaN        NaN        NaN      NaN   
5               NaN              NaN        NaN        NaN      NaN   

          Q10_Part_1       Q10_Part_2 Q10_Part_3 Q10_Part_4 Q10_Part_5  \
1   Kaggle Notebooks  Colab Notebooks        NaN        NaN        NaN   
2                NaN  Colab Notebooks        NaN        NaN        NaN   
3                NaN              NaN        NaN        NaN        NaN   
4                NaN              NaN        NaN        NaN        NaN   
5                NaN              NaN        NaN        NaN        NaN   

  Q10_Part_6 Q10_Part_7 Q10_Part_8 Q10_Part_9 Q10_Part_10 Q10_Part_11  \
1        NaN        NaN        NaN        NaN         NaN         NaN   
2        NaN        NaN        NaN        NaN         NaN         NaN   
3        NaN        NaN        NaN        NaN         NaN         NaN   
4        NaN        NaN        NaN        NaN         NaN         NaN   
5        NaN        NaN        NaN        NaN         NaN         NaN   

  Q10_Part_12 Q10_Part_13 Q10_OTHER  \
1         NaN         NaN     

In [8]:
#Remove the first row 
df= df.iloc[1:]

In [9]:
#Let's take only in consideration Data Scientist and Data Analyst
df=df.loc[(df.Q5=='Data Scientist') | (df.Q5=='Data Analyst')]

In [176]:
import plotly.graph_objects as go

#color used (documentation plotly)
colors = ['lightslategray',] * 5
colors[1] = 'crimson'

#Define our x and y
x = df['Q5']
y = (round(df['Q5'].value_counts(normalize=True),2)*100)

# Use the hovertext kw argument for hover text
fig = go.Figure(data=[go.Bar(x=x, y=y, text=y,textposition='auto',
            hovertext=['64% of the dataset', '36% of the dataset'])]
)
# Customize aspect
fig.update_traces(marker_color=colors,
                  marker_line_width=1., opacity=0.6)
fig.update_layout(title_text="<span style='font-size:36px; font-family:Times New Roman'>Proportion of Data Analyst and Data Scientist</span>", width=800
    ,height=500,)

#Annotation
fig.add_annotation(dict(
        x=0.3,
        y=ds_degree.values[0] + 10,
        text="<b>2/3 our data are from Data Scientist",
        ax="38",
        ay="-50",
        showarrow = True,
        arrowhead = 7,
        arrowwidth = 0.7
    
))

fig.show()

In [11]:
#We will have two datasets, one for data analyst the other for da
ds=df.loc[(df.Q5=='Data Scientist')]
da=df.loc[(df.Q5=='Data Analyst')]

In [12]:
#Data scientist general statistics
ds.describe()

Time from Start to Finish (seconds)     Q1    Q2     Q3  \
count                                  2676   2676  2676   2676   
unique                                 1604     11     5     55   
top                                     469  25-29   Man  India   
freq                                     11    673  2194    583   

                     Q4              Q5         Q6 Q7_Part_1 Q7_Part_2  \
count              2676            2676       2659      2453       975   
unique                7               1          7         1         1   
top     Master’s degree  Data Scientist  3-5 years    Python         R   
freq               1369            2676        743      2453       975   

       Q7_Part_3 Q7_Part_4 Q7_Part_5 Q7_Part_6   Q7_Part_7 Q7_Part_8  \
count       1473       247       323       305         290        71   
unique         1         1         1         1           1         1   
top          SQL         C       C++      Java  Javascript     Julia   
freq        1473       247       323       305         290        71   

       Q7_Part_9 Q7_Part_10 Q7_Part_11 Q7_Part_12 Q7_OTHER      Q8  \
count         14        405        242          5      245    2606   
unique         1          1          1          1        1      13   
top        Swift       Bash     MATLAB       None    Other  Python   
freq          14        405        242          5      245    2087   

                                            Q9_Part_1  Q9_Part_2  \
count                                            1918        814   
unique                                              1          1   
top     Jupyter (JupyterLab, Jupyter Notebooks, etc)    RStudio    
freq                                             1918        814   

            Q9_Part_3                    Q9_Part_4  Q9_Part_5   Q9_Part_6  \
count             262                          860        825         564   
unique              1                            1          1           1   
top     Visual Studio  Visual Studio Code (VSCode)   PyCharm     Spyder     
freq              262                          860        825         564   

            Q9_Part_7         Q9_Part_8        Q9_Part_9 Q9_Part_10  \
count             504               393              285        151   
unique              1                 1                1          1   
top       Notepad++      Sublime Text      Vim / Emacs      MATLAB    
freq              504               393              285        151   

       Q9_Part_11 Q9_OTHER         Q10_Part_1       Q10_Part_2  \
count          17      146                917             1058   
unique          1        1                  1                1   
top          None    Other   Kaggle Notebooks  Colab Notebooks   
freq           17      146                917             1058   

             Q10_Part_3               Q10_Part_4             Q10_Part_5  \
count               178                       35                    326   
unique                1                        1                      1   
top     Azure Notebooks   Paperspace / Gradient    Binder / JupyterHub    
freq                178                       35                    326   

          Q10_Part_6           Q10_Part_7                 Q10_Part_8  \
count             12                  118                        162   
unique             1                    1                          1   
top      Code Ocean    IBM Watson Studio    Amazon Sagemaker Studio    
freq              12                  118                        162   

                    Q10_Part_9                          Q10_Part_10  \
count                       55                                  223   
unique                       1                                    1   
top      Amazon EMR Notebooks   Google Cloud AI Platform Notebooks    
freq                        55                                  223   

                           Q10_Part_11                           Q10_Part_12  \
co

Our typical data scientist is between 25 and 29 years old, have between 3 and 5 years expriences. He is man from India who use Python as regular basis.

In [13]:
da.describe()

Time from Start to Finish (seconds)     Q1    Q2     Q3  \
count                                  1475   1475  1475   1475   
unique                                 1086     11     5     55   
top                                     659  25-29   Man  India   
freq                                      8    391  1110    343   

                     Q4            Q5         Q6 Q7_Part_1 Q7_Part_2  \
count              1475          1475       1457      1123       471   
unique                7             1          7         1         1   
top     Master’s degree  Data Analyst  1-2 years    Python         R   
freq                715          1475        423      1123       471   

       Q7_Part_3 Q7_Part_4 Q7_Part_5 Q7_Part_6   Q7_Part_7 Q7_Part_8  \
count        815       111       143       107         147         9   
unique         1         1         1         1           1         1   
top          SQL         C       C++      Java  Javascript     Julia   
freq         815       111       143       107         147         9   

       Q7_Part_9 Q7_Part_10 Q7_Part_11 Q7_Part_12 Q7_OTHER      Q8  \
count          9         74        105         12      128    1353   
unique         1          1          1          1        1      12   
top        Swift       Bash     MATLAB       None    Other  Python   
freq           9         74        105         12      128     977   

                                            Q9_Part_1  Q9_Part_2  \
count                                             870        417   
unique                                              1          1   
top     Jupyter (JupyterLab, Jupyter Notebooks, etc)    RStudio    
freq                                              870        417   

            Q9_Part_3                    Q9_Part_4  Q9_Part_5   Q9_Part_6  \
count             145                          323        305         259   
unique              1                            1          1           1   
top     Visual Studio  Visual Studio Code (VSCode)   PyCharm     Spyder     
freq              145                          323        305         259   

            Q9_Part_7         Q9_Part_8        Q9_Part_9 Q9_Part_10  \
count             252               156               52         74   
unique              1                 1                1          1   
top       Notepad++      Sublime Text      Vim / Emacs      MATLAB    
freq              252               156               52         74   

       Q9_Part_11 Q9_OTHER         Q10_Part_1       Q10_Part_2  \
count          35       79                421              369   
unique          1        1                  1                1   
top          None    Other   Kaggle Notebooks  Colab Notebooks   
freq           35       79                421              369   

             Q10_Part_3               Q10_Part_4             Q10_Part_5  \
count                74                        9                    167   
unique                1                        1                      1   
top     Azure Notebooks   Paperspace / Gradient    Binder / JupyterHub    
freq                 74                        9                    167   

          Q10_Part_6           Q10_Part_7                 Q10_Part_8  \
count             14                   80                         25   
unique             1                    1                          1   
top      Code Ocean    IBM Watson Studio    Amazon Sagemaker Studio    
freq              14                   80                         25   

                    Q10_Part_9                          Q10_Part_10  \
count                       25                                   84   
unique                       1                                    1   
top      Amazon EMR Notebooks   Google Cloud AI Platform Notebooks    
freq                        25                                   84   

                           Q10_Part_11                           Q10_Part_12  \
count       

## Data Scientist VS Data Analyst: Age distribution 

In [55]:
#let's create some color for the rest of the analysis
primary_blue = "#496595"
primary_grey = "#c6ccd8"
primary_salmon="#cd5c5c"
primary_seashell="#FFF5EE"

In [143]:
ds=df.loc[(df.Q5=='Data Scientist')]
da=df.loc[(df.Q5=='Data Analyst')]

ds_ages = round(ds["Q1"].value_counts(normalize=True),2)
da_ages = round(da["Q1"].value_counts(normalize=True),2)

trace1 = go.Bar(
    x = ds_ages.keys(),
    y = ds_ages.values,
    name = "Data Scientist",
    marker_color = primary_salmon,
    text = ds_ages[ds_ages.keys()],
    textposition = "outside",
#     width = 0.6
)

trace2 = go.Bar(
    x = ds_ages.keys(),
    y = - da_ages[ds_ages.keys()],
    name = "Data Analyst",
    marker_color = primary_seashell,
    text = da_ages[ds_ages.keys()],
    textposition = "outside",
)

layout = dict(
    title = "<span style='font-size:36px; font-family:Times New Roman'>Age proportion by position</span>",
    plot_bgcolor='#fff',
    margin = dict(t=50, l=0, r=0),
    legend=dict(title="  Role:", yanchor='top',xanchor='right', x=1.0, y=1.0, font=dict(family="Tahoma", size= 14),traceorder='normal', bordercolor="#4d4d4d", borderwidth=0.5),
    xaxis = dict(domain=[0,1]),
    yaxis = dict(showticklabels = False, zeroline=True, zerolinecolor="#4d4d4d", zerolinewidth=0.5, showgrid=False),
    barmode="overlay",
    bargap = 0.1,
    width = 765
)



data = [trace1, trace2]
fig = go.Figure(data = data)

fig['layout'].update(layout)
fig.update_xaxes(categoryarray=["18-21", "22-24", "25-29", "30-34","35-39","40-44","45-49","50-54","55-59","60-69","70+"])
# Adding anotations

fig.add_annotation(dict(
        x=2.0,
        y=0.4,
        text="<b>More than 50%</b>of the data scientist<br> and data analyst respondents<br>are between 22 and 34 years old.",
        ax="38",
        ay="-50",
        showarrow = False,
        arrowhead = 7,
        arrowwidth = 0.7
    
))


# Rectangle to highlight range
fig.add_vrect(x0=0.5, x1=3.47,
              fillcolor="#D3D3D3",
              layer="below", 
              opacity=0.25, 
              line_width=0
)

iplot(fig)

In [181]:
ds=df.loc[(df.Q5=='Data Scientist')]
da=df.loc[(df.Q5=='Data Analyst')]




ds_degree = round(ds["Q4"].value_counts(normalize=True),2)*100
da_degree = round(da["Q4"].value_counts(normalize=True),2)*100

trace1 = go.Bar(
    x = ds_degree.keys(),
    y = ds_degree.values,
    name = "Data Scientist",
    marker_color = primary_blue,
    text = ds_degree[ds_degree.keys()],
    textposition = "outside",
#     width = 0.6
)

trace2 = go.Bar(
    x = ds_degree.keys(),
    y = - da_degree[ds_degree.keys()],
    name = "Data Analyst",
    marker_color = primary_grey,
    text = da_degree[ds_degree.keys()],
    textposition = "outside",
)

layout = dict(
    title = "<span style='font-size:36px; font-family:Times New Roman'>Percentage of respondents graduated per position</span>",
    plot_bgcolor='#fff',
    margin = dict(t=50, l=0, r=0),
    legend=dict(title="  Legend:", yanchor='top',xanchor='right', x=1.15, y=1, font=dict(family="Tahoma", size= 14),traceorder='normal', bordercolor="#4d4d4d", borderwidth=0.5),
    xaxis = dict(domain=[0,1]),
    yaxis = dict(showticklabels = False, zeroline=True, zerolinecolor="#4d4d4d", zerolinewidth=0.5, showgrid=False),
    barmode="overlay",
    bargap = 0.1,
    width = 765
)



data = [trace1,trace2]
fig = go.Figure(data = data)

fig['layout'].update(layout)

# Adding anotations
fig.add_annotation(dict(
        x=0.3,
        y=ds_degree.values[0] + 10,
        text="Approximately <b>50% of the Data Analyst<br>or Data Scientist</b> respondents<br>have a <b>master's degrees.</b>",
        ax="38",
        ay="-50",
        showarrow = True,
        arrowhead = 7,
        arrowwidth = 0.7
    
))

fig.add_annotation(dict(
        x=2.1,
        y=ds_degree.values[0] + 5,
        text="We can see a <br> <b>difference for Phd <br> and bachelor's degree.</b>",
        showarrow = False
    
))



# Rectangle to highlight range
fig.add_vrect(x0=2.53, x1=5.52,
              fillcolor="#e9cf87",
              layer="below", 
              opacity=0.25, 
              line_width=0
)





fig.add_annotation(dict(
        x=4.03,
        y=ds_degree[ds_degree.keys()][1] +50,
        text="A very <b> small proportion of data <br> analyst or scientist</b> has <br> a degree <b>under bachelor.<b/>",
        showarrow = False
))

iplot(fig)

In [18]:
#create a dictionnary for question
Questions={}
#create list of questions 
#not very efficient, but keeps things ordered
qnums=list(dict.fromkeys([i.split('_')[0] for i in ds.columns]))
qnums

['Time from Start to Finish (seconds)',
 'Q1',
 'Q2',
 'Q3',
 'Q4',
 'Q5',
 'Q6',
 'Q7',
 'Q8',
 'Q9',
 'Q10',
 'Q11',
 'Q12',
 'Q13',
 'Q14',
 'Q15',
 'Q16',
 'Q17',
 'Q18',
 'Q19',
 'Q20',
 'Q21',
 'Q22',
 'Q23',
 'Q24',
 'Q25',
 'Q26',
 'Q27',
 'Q28',
 'Q29',
 'Q30',
 'Q31',
 'Q32',
 'Q33',
 'Q34',
 'Q35',
 'Q36',
 'Q37',
 'Q38',
 'Q39']

In [19]:
def get_percent(x):
    
    for i in qnums:
        if i in ['Q1', 'Q2', 'Q3']:
            Questions[i]=x[i]
        else:
            Questions[i]= x[[q for q in da.columns if q.startswith(i)]]
        
    Questions['Q23'].columns = list(Questions['Q23'].mode().values)


    result = Questions['Q23'].count().reset_index()
    result.columns=['Activities', 'Count']
    result=result.sort_values(by=['Count'], ascending=False)

    #result=result.loc[q23.Activities!='Other']
    result['total']=result['Count'].sum()
    result['percentage']=round(result['Count']/result['total'],2)*100
    result=result[['Activities', 'percentage']]
    
    return result

In [20]:
da_per=get_percent(da)
ds_per=get_percent(ds)

In [21]:
da_per

Activities  percentage
0  Analyze and understand data to influence produ...        43.0
1  Build and/or run the data infrastructure that ...        18.0
2  Build prototypes to explore applying machine l...        11.0
4  Experimentation and iteration to improve exist...         8.0
3  Build and/or run a machine learning service th...         8.0
5  Do research that advances the state of the art...         6.0
6  None of these activities are an important part...         3.0
7                                              Other         3.0

In [188]:
import plotly.express as px
fig = px.pie(da_per, values='percentage', names=[
    'Analyze data for business decisions',
    'Build the data infrastructure',
    'Build ML prototypes in new areas',
    'Build ML services for products/workflows',
    'Improving existing ML models',
    'Research to advance ML',
    'None','Other'], color_discrete_sequence=px.colors.sequential.RdBu, title="The main acitivities of the data analyst")

fig.add_annotation(x=-0.05, y=1,
            text="More than <b>40% of the activities</b><br> of the Data Analyst on analyzing<br> on <b>business decision</b>",
            showarrow=False,
            arrowhead=1)

fig.add_annotation(x=-0.05, y=0,
            text="We can see the <b>machine<br> learning is also very common <b/> <br>for a Data Analyst position",
            showarrow=False,
            arrowhead=1)

fig.show()

In [185]:
import plotly.express as px
fig = px.pie(ds_per, values='percentage', names=[
    'Analyze data for business decisions',
    'Build the data infrastructure',
    'Build ML prototypes in new areas',
    'Build ML services for products/workflows',
    'Improving existing ML models',
    'Research to advance ML',
    'None','Other'], color_discrete_sequence=px.colors.sequential.RdBu, title="The main activities for a data scientist")

fig.add_annotation(x=-0.05, y=1,
            text="<b>43%</b> of the activities are<br> related to <b>Machine Learning</b>",
            showarrow=False,
            arrowhead=1)

fig.add_annotation(x=-0.05, y=0,
            text="<b>46%</b> of the tasks  are<br> focused on building the <b>data<br>infrastructure</b> and <b>business <br>decision<b/>",
            showarrow=False,
            arrowhead=1)

fig.show()

# Programming language

In [26]:
def language_percent(x):
    
    for i in qnums:
        if i in ['Q1', 'Q2', 'Q3']:
            Questions[i]=x[i]
        else:
            Questions[i]= x[[q for q in da.columns if q.startswith(i)]]
        
    Questions['Q7'].columns = list(Questions['Q7'].mode().values)


    result = Questions['Q7'].count().reset_index()
    result.columns=['Language', 'Count']
    result=result.sort_values(by=['Count'], ascending=False)

    #result=result.loc[q23.Activities!='Other']
    result['total']=result['Count'].sum()
    result['percentage']=round(result['Count']/result['total'],2)*100
    result=result[['Language', 'percentage']]
    
    return result

In [27]:
ds_language=language_percent(ds)
da_language=language_percent(da)

In [28]:
ds_language

Language  percentage
0       Python        35.0
2          SQL        21.0
1            R        14.0
9         Bash         6.0
4          C++         5.0
5         Java         4.0
6   Javascript         4.0
3            C         4.0
12       Other         3.0
10      MATLAB         3.0
7        Julia         1.0
8        Swift         0.0
11        None         0.0

In [29]:
da_language

Language  percentage
0       Python        35.0
2          SQL        25.0
1            R        14.0
6   Javascript         5.0
4          C++         4.0
12       Other         4.0
3            C         3.0
5         Java         3.0
10      MATLAB         3.0
9         Bash         2.0
11        None         0.0
7        Julia         0.0
8        Swift         0.0

In [30]:
import plotly.express as px
fig = px.pie(da_language, values='percentage', names=da_language['Language'], color_discrete_sequence=px.colors.sequential.RdBu, title="Most used languages for a data analayst")

fig.add_annotation(x=-0.05, y=1,
            text="<b>43%</b> of the activities are<br> related to <b>Machine Learning</b>",
            showarrow=False,
            arrowhead=1)

fig.add_annotation(x=-0.05, y=0,
            text="<b>46%</b> of the tasks  are<br> focused on building the <b>data<br>infrastructure</b> and <b>business decision<b/>",
            showarrow=False,
            arrowhead=1)

fig.show()

In [31]:
import plotly.express as px
fig = px.pie(ds_language, values='percentage', names=ds_language['Language'], color_discrete_sequence=px.colors.sequential.RdBu, title="Most common languages for a Data Scientist")

fig.add_annotation(x=-0.05, y=1,
            text="<b>43%</b> of the activities are<br> related to <b>Machine Learning</b>",
            showarrow=False,
            arrowhead=1)

fig.add_annotation(x=-0.05, y=0,
            text="<b>46%</b> of the tasks  are<br> focused on building the <b>data<br>infrastructure</b> and <b>business decision<b/>",
            showarrow=False,
            arrowhead=1)

fig.show()

In [32]:
ds_language.head()

Language  percentage
0   Python        35.0
2      SQL        21.0
1        R        14.0
9     Bash         6.0
4      C++         5.0

In [33]:
da_language.head()

Language  percentage
0      Python        35.0
2         SQL        25.0
1           R        14.0
6  Javascript         5.0
4         C++         4.0

In [187]:
categories = ['Python','SQL', 'R', 'Javascript','C++','Bash']

fig = go.Figure()

fig.add_trace(go.Scatterpolar(
      r=ds_language['percentage'],
      theta=categories,
      fill='toself',
      name='Data Scientist'
))
fig.add_trace(go.Scatterpolar(
      r=da_language['percentage'],
      theta=categories,
      fill='toself',
      name='Data Analyst'
))

fig.add_annotation(x=-0.0, y=0.9,
            text="<b>35%</b> of the respondents <br>Data Scientist and Data Analyst<br> used <b>Python<b/> as regular basis",
            showarrow=False,
            arrowhead=1)

fig.add_annotation(x=-0.0, y=0.1,
            text="<b>1/4 of the Data Analysis</b> <br>of the study declared using<br> <b> SQL as regular basis</b>",
            showarrow=False,
            arrowhead=1)



fig.update_layout(
    title={
        'text': "<span style='font-size:36px; font-family:Times New Roman'>Data Scientist VS Data Analyst: Most common languages</span>",
        'y':0.9,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
    polar=dict(
    radialaxis=dict(
      visible=True,
      range=[0, 35]
    )),
  showlegend=True
)



fig.show()


# Machine Learning Usage 

## ML librairies used as regular basis

In [109]:
def librairie_ml(x):
    
    for i in qnums:
        if i in ['Q1', 'Q2', 'Q3']:
            Questions[i]=x[i]
        else:
            Questions[i]= x[[q for q in da.columns if q.startswith(i)]]
        
    Questions['Q16'].columns = list(Questions['Q16'].mode().values)


    result = Questions['Q16'].count().reset_index()
    result.columns=['Librairies', 'Count']
    result=result.sort_values(by=['Count'], ascending=False)

    #result=result.loc[q23.Activities!='Other']
    result['total']=result['Count'].sum()
    result['percentage']=round(result['Count']/result['total'],2)*100
    result=result[['Librairies', 'percentage']]
    
    return result

In [113]:
ds_ml_librairies=librairie_ml(ds)
da_ml_librairies=librairie_ml(da)

In [114]:
ds_ml_librairies

Librairies  percentage
0     Scikit-learn         23.0
2            Keras         14.0
1       TensorFlow         14.0
6          Xgboost         14.0
3          PyTorch          9.0
7         LightGBM          7.0
11           Caret          4.0
8         CatBoost          4.0
9          Prophet          3.0
4          Fast.ai          2.0
12      Tidymodels          2.0
10           H2O 3          2.0
15            Other         1.0
14             None         1.0
5            MXNet          1.0
13             JAX          0.0

In [115]:
da_ml_librairies

Librairies  percentage
0     Scikit-learn         29.0
1       TensorFlow         15.0
2            Keras         13.0
6          Xgboost         11.0
3          PyTorch          9.0
14             None         5.0
7         LightGBM          4.0
11           Caret          4.0
12      Tidymodels          2.0
8         CatBoost          2.0
9          Prophet          2.0
10           H2O 3          1.0
4          Fast.ai          1.0
15            Other         1.0
13             JAX          0.0
5            MXNet          0.0

categories = ['Scikit-learn','Keras', 'TensorFlow', 'Xgboost','PyTorch']

fig = go.Figure()

fig.add_trace(go.Scatterpolar(
      r=ds_ml_librairies['percentage'],
      theta=categories,
      fill='toself',
      name='Data Scientist'
))
fig.add_trace(go.Scatterpolar(
      r=da_ml_librairies['percentage'],
      theta=categories,
      fill='toself',
      name='Data Analyst'
))

fig.add_annotation(x=-0.075, y=0.93,
            text="<b>The 5 most popular ML librairies</b> <br>used as regular basis <b>are the same <b/><br> between Data Analyst <br>and Data Scientist",
            showarrow=False,
            arrowhead=1)

fig.add_annotation(x=-0.14, y=0.1,
            text="<b>Surprisingly the deep learning</b> <br>librairies <b>Keras,Tensorflow and pytorch</b> <br>are as common for <b>Data Scientist <br>and Data analyst</b>",
            showarrow=False,
            arrowhead=1)



fig.update_layout(
    title={
        'text': "<span style='font-size:36px; font-family:Times New Roman'>Most popular ML librairies used as regular basis per role</span>",
        'y':0.9,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'},
    polar=dict(
    radialaxis=dict(
      visible=True,
      range=[0, 35]
    )),
  showlegend=True
)



fig.show()
